frame/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,228 @@
+		type Staking: StakingInterface<AccountId = Self::AccountId>;
+
+		/// A sorted list of nominators and validators, by their stake and self-stake respectively.
+		type VoterList: SortedListProvider<Self::AccountId, Score = VoteWeight>;
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	pub(crate) fn active_stake_of(who: &T::AccountId) -> BalanceOf<T> {
+		T::Staking::stake(&who).map(|s| s.active).unwrap_or_default()
+	}
+
+	pub(crate) fn to_vote(balance: BalanceOf<T>) -> VoteWeight {
+		let total_issuance = T::Currency::total_issuance();
+		<T::Staking as StakingInterface>::CurrencyToVote::to_vote(balance, total_issuance)
+	}
+}
+
+impl<T: Config> OnStakingUpdate<T::AccountId, BalanceOf<T>> for Pallet<T> {
+	fn on_stake_update(who: &T::AccountId, _: Option<Stake<T::AccountId, BalanceOf<T>>>) {
+		if let Ok(current_stake) = T::Staking::stake(who) {
+			let current_active = current_stake.active;
+
+			// If this is a nominator, update their position in the `VoterList`.
+			if let Some(_) = T::Staking::nominations(&current_stake.stash) {
+				let _ =
+					T::VoterList::on_update(&current_stake.stash, Self::to_vote(current_active))
+						.defensive_proof("Nominator's position in VoterList updated; qed");
+			}
+
+			// If this is a validator, update their position in the `VoterList`.
+			if T::Staking::is_validator(&current_stake.stash) {
+				let _ =
+					T::VoterList::on_update(&current_stake.stash, Self::to_vote(current_active))
+						.defensive_proof("Validator's position in VoterList updated; qed");
+			}
+		}
+	}
+	fn on_nominator_add(who: &T::AccountId) {
+		let _ = T::VoterList::on_insert(who.clone(), Self::to_vote(Self::active_stake_of(who)))
+			.defensive_proof("Nominator inserted into VoterList; qed");
+	}
+
+	fn on_nominator_update(who: &T::AccountId, _prev_nominations: Vec<T::AccountId>) {
+		if !T::VoterList::contains(who) {
+			defensive!("Active nominator is in the VoterList; qed");
+		}
+	}
+
+	fn on_validator_add(who: &T::AccountId) {
+		let _ = T::VoterList::on_insert(who.clone(), Self::to_vote(Self::active_stake_of(who)))
+			.defensive_proof("Validator inserted into VoterList; qed");
+	}
+
+	fn on_validator_update(who: &T::AccountId) {
+		if !T::VoterList::contains(who) {
+			defensive!("Active validator is in the VoterList; qed");
+		}
+	}
+
+	fn on_validator_remove(who: &T::AccountId) {
+		let _ =
+			T::VoterList::on_remove(who).defensive_proof("Validator removed from VoterList; qed");
+	}
+
+	fn on_nominator_remove(who: &T::AccountId, _nominations: Vec<T::AccountId>) {
+		let _ =
+			T::VoterList::on_remove(who).defensive_proof("Nominator removed from VoterList; qed");
+	}
+
+	fn on_unstake(who: &T::AccountId) {
+		if T::VoterList::contains(who) {
+			defensive!("The staker has already been removed; qed");
+		}
+	}
+}
+
+/// A wrapper for a given `SortedListProvider` that introduces defensive checks  for insert, update
+/// and remove operations, suggesting that it's read-only, except for unsafe operations.
```

Comment:
Maybe you could elaborate more what are unsafe operations and why we need them here? It sounds important